### PR TITLE
Add tests for OVS QoS bandwidth limitations

### PR DESCRIPTION
**Pull request by oribon**
These tests verify that egress/ingress limitation applied by ovs QoS MachineConfigs
work properly

### CODE REVIEW

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+	if os.Getenv("QOS_TEST_EGRESS") == "false" {
+		testEgress = false
+	}
+
+	testIngress = true
+	if os.Getenv("QOS_TEST_INGRESS") == "false" {
+		testIngress = false
+	}
+
+	egressOverride = os.Getenv("QOS_TEST_EGRESS_OVERRIDE")
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:71]: is it possible to be dynamic here? start a pods with iperf get the numbers apply and the validate again with pods?

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 2]
```diff
@@ -0,0 +1,644 @@
...
+		}
+		if !discovery.Enabled() {
+			if roleWorkerCNF == "" {
+				gSkipReason = "ROLE_WORKER_CNF was not specified, can't determine nodes for ovs_qos MachineConfig without discovery."
+				return
+			}
+		}
+	})
+
+	Context("Test egress", func() {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:101]: maybe "validate egress QOS limitation`

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+		var receiverNode string
+		var senderNode string
+		mcpUpdated := false
+		var err error
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:105]: nit: change to 

```golang
var receiverNode string
var senderNode string
var err error
var mcpName string

eSkipReason := ""
mcpUpdated := false
```

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 4]
```diff
@@ -0,0 +1,644 @@
...
+		eSkipReason := ""
+		var mcpName string
+		execute.BeforeAll(func() {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:108]: I think you can have a bool if the machine config exist and remove the beforeAll just have the beforeEach

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 5]
```diff
@@ -0,0 +1,644 @@
...
+				Skip("No egress MC was applied by the test")
+			}
+			err = client.Client.MachineConfigs().Delete(context.Background(), egressMCName, metav1.DeleteOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			err = waitForMCPStable(mcpName)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+	})
+	Context("Test ingress", func() {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:207]: maybe change to `validate egress QOS limitation`

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 6]
```diff
@@ -0,0 +1,644 @@
...
+		var receiverNode string
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:208]: same comments as before both for variables and the beforeAll

**Review by SchSeba — COMMENTED** [review]
looks good just some comments I think this PR is in the right direction

**Comment by SchSeba** [comment]
@oribon please rebase this PR

**Comment by fedepaol** [comment]
I'd rather add the MC deployment as part of the configuration and treat it as a feature, instead of creating the MC in the test.

**Comment by SchSeba** [comment]
@fedepaol I was thinking that will be the case in discovery mode. in the regular mode was to apply the policy and test it.

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 7]
```diff
@@ -0,0 +1,989 @@
...
+
+				By("Creating receiver pod")
+				receiverPod := createReceiverPod(receiverNode, receiverHostNetwork)
+				receiverIP := receiverPod.Status.PodIP
+
+				By("Creating sender pod")
+				senderPod := createSenderPod(senderNode, senderHostNetwork, receiverIP)
+
+				By("Waiting for the sender to end")
+				time.Sleep(15 * time.Second)
```
Comment (kononovn) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:219]: Can we add here Eventually loop which will wait until sender pod is in Completed state? In additional we has to remove "Sleep INF" command from sender pod

Comment (oribon) [reply cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go]: done

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 8]
```diff
@@ -0,0 +1,980 @@
...
+
+					if mcpUpdated {
+						err = waitForMCPStable(mcpName)
+						if err != nil {
+							eSkipReason = err.Error()
+							Fail(err.Error())
+						}
+					}
+				}
+
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:187]: nit: remove empty line

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 9]
```diff
@@ -0,0 +1,980 @@
...
+				return fmt.Errorf("egress QoS not applied correctly on node %s for limit: %d", node.Name, limit)
+			}
+		} else {
+			matches = findLimit.FindStringSubmatch(qos)
+			if len(matches) < 2 {
+				return fmt.Errorf("couldn't find rate from ovs-appctl qos/show on node %s", node.Name)
+			}
+
+			appliedLimit, err = strconv.ParseInt(matches[1], 10, 64)
+
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:935]: nit: remove empty line

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 10]
```diff
@@ -0,0 +1,980 @@
...
+func findQoSRateWithNodeSelector(qosType func(string) float64, qosNodeSelector string) (string, float64, error) {
+	mcList, err := client.Client.MachineConfigs().List(context.Background(), metav1.ListOptions{})
+	Expect(err).ToNot(HaveOccurred())
+
+	for _, mc := range mcList.Items {
+		enables, err := isMCLimiting(mc, qosType)
+		if err != nil {
+			return "", -1, err
+		}
+		if enables != -1 {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:523]: nit: change to enables == -1 return error

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 11]
```diff
@@ -0,0 +1,980 @@
...
+	}
+	bitRate, err := strconv.ParseFloat(m[1], 64)
+	if err != nil {
+		return -1
+	}
+	// ingress QoS is set in kbits
+	return bitRate * 1000
+}
+
+func findQoSNodeSelectorByMCLabel(mcLabel string) (string, error) {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:591]: we have this already multiple times in the code

please lets make this https://github.com/openshift-kni/cnf-features-deploy/blob/master/cnf-tests/testsuites/validationsuite/cluster/validation.go#L416

to be under the pkg/machineconfig so all the tests can reuse it.

follow up PR do the same change for x32 and sctp tests

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 12]
```diff
@@ -0,0 +1,980 @@
...
+					return newQoSNodeSelector, nil
+				}
+			}
+		}
+	}
+
+	return "", fmt.Errorf("cannot find QoSNodeSelector")
+}
+
+func findMCPByMCRole(mcLabel string) (string, error) {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:619]: ditto

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 13]
```diff
@@ -0,0 +1,980 @@
...
+func rateEnforced(o iperf3Output, limitRate float64) bool {
+	for _, sum := range o.Intervals {
+		if sum.Sum.BitsPerSecond > 1.05*limitRate {
+			return false
+		}
+	}
+	return true
+}
+
+func waitForMCPStable(mcpName string) error {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:649]: this funtion already exist under pkg/machineconfig

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 14]
```diff
@@ -0,0 +1,980 @@
...
+		client.Client,
+		&mcfgv1.MachineConfigPool{ObjectMeta: metav1.ObjectMeta{Name: mcpName}},
+		mcfgv1.MachineConfigPoolUpdated,
+		corev1.ConditionTrue,
+		time.Duration(30*mcp.Status.MachineCount)*time.Minute)
+
+	return err
+}
+
+func createEgressMC(egressRate int64) (*mcfgv1.MachineConfig, error) {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:679]: please have a template for this so you can use it for both Egress and Ingress

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 15]
```diff
@@ -0,0 +1,980 @@
...
+	if rate == -1 {
+		return nil, fmt.Errorf("egress limit requested is not valid")
+	}
+	err = client.Client.Create(context.TODO(), mc)
+	return mc, err
+}
+
+func createIngressMC(ingressRate int64) (*mcfgv1.MachineConfig, error) {
+
+	mcContent := fmt.Sprintf(`
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:734]: ditto

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 16]
```diff
@@ -0,0 +1,980 @@
...
+}
+
+func decodeMCYaml(mcyaml string) (*mcfgv1.MachineConfig, error) {
+	decode := mcfgScheme.Codecs.UniversalDeserializer().Decode
+	obj, _, err := decode([]byte(mcyaml), nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	mc, ok := obj.(*mcfgv1.MachineConfig)
+	Expect(ok).To(Equal(true))
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:785]: better to have if here don't use both return error and ginkgo fundtions

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 17]
```diff
@@ -0,0 +1,980 @@
...
+	s = s / float64(len(o.Intervals))
+
+	if s == 0 {
+		return 0, fmt.Errorf("rate sum of iperf3 intervals is 0, can't determine rate")
+	}
+
+	return s, nil
+}
+
+func egressQosOnNode(node *corev1.Node) (string, error) {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:828]: please make a reuse fection where you just send the command

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 18]
```diff
@@ -0,0 +1,980 @@
...
+
+	buf, err := pods.ExecCommand(client.Client, *ovsPod, []string{"/bin/bash", "-c", fmt.Sprintf("/usr/bin/ovs-appctl qos/show %s", nic)})
+	if err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+func ingressQosOnNode(node *corev1.Node) (string, error) {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:847]: ditto

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 19]
```diff
@@ -0,0 +1,980 @@
...
+	return buf.String(), nil
+}
+
+func findOvsPhysicalNic(ovsPod corev1.Pod) (string, error) {
+	buf, err := pods.ExecCommand(client.Client, ovsPod, []string{"/bin/bash", "-c", "/usr/bin/ovs-vsctl list-ports br-ex | /usr/bin/grep -v patch"})
+	if err != nil {
+		return "", err
+	}
+
+	return regexp.MustCompile("\r\n").ReplaceAllString(buf.String(), ""), nil
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:872]: maybe check if it's empty here and return error?

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 20]
```diff
@@ -0,0 +1,980 @@
...
+
+		if !strings.Contains(qos, fmt.Sprintf("ingress_policing_rate: %d", limit)) {
+			return fmt.Errorf("ingress QoS not applied correctly on node %s for limit: %d", node.Name, limit)
+		}
+	}
+
+	return nil
+}
+
+func validateEgressQosOnNodes(limit int64) error {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:902]: I think is possible to create a combine function for both egress and ingress

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 21]
```diff
@@ -0,0 +1,980 @@
...
+	err = pods.WaitForCondition(client.Client, receiverPod, corev1.ContainersReady, corev1.ConditionTrue, 2*time.Minute)
+	Expect(err).ToNot(HaveOccurred())
+
+	By("Getting updated receiver spec")
+	receiverPod, err = client.Client.Pods(receiverPod.Namespace).Get(context.Background(), receiverPod.Name, metav1.GetOptions{})
+	Expect(err).ToNot(HaveOccurred())
+	return receiverPod
+}
+
+func defineReceiverPod(nodeName string, hostNetwork bool) *corev1.Pod {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:490]: combine and just send the command

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 22]
```diff
@@ -0,0 +1,980 @@
...
+		pod = pods.DefinePodOnHostNetwork(namespaces.OVSQOSTest, nodeName)
+	} else {
+		pod = pods.DefinePodOnNode(namespaces.OVSQOSTest, nodeName)
+	}
+	pod = pods.RedefineWithCommand(pod, []string{"/bin/bash", "-c"}, []string{"iperf3 -s -J -1;"})
+	pod = pods.RedefineWithRestartPolicy(pod, corev1.RestartPolicyNever)
+	return pod
+}
+
+func defineSenderPod(nodeName string, hostNetwork bool, receiverIP string) *corev1.Pod {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:502]: ditto

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 23]
```diff
@@ -0,0 +1,980 @@
...
+	}
+
+	if len(parsedOutput.Intervals) == 0 {
+		return parsedOutput, fmt.Errorf("parsed iperf3output returned 0 intervals")
+	}
+
+	return parsedOutput, nil
+}
+
+func createSenderPod(nodeName string, hostNetwork bool, receiverIP string) *corev1.Pod {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:462]: I think the two functions can be combine no?

**Review by SchSeba — COMMENTED** [review]
Thanks for the PR!

I just add some comments from an initial review

File: cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go [hunk 24]
```diff
@@ -0,0 +1,871 @@
...
+
+	buf, err := pods.ExecCommand(client.Client, *ovsPod, []string{"/bin/bash", "-c", fmt.Sprintf(cmdWithNICTemplate, nic)})
+	if err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+func findOvsPhysicalNic(ovsPod corev1.Pod) (string, error) {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:748]: input variable for pod should be a pointer

Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go:748]: please add an example output and explain on the function it will be really hard to debug it

**Review by SchSeba — COMMENTED** [review]
almost there nice work!

**Comment by SchSeba** [comment]
@oribon the PR looks good please add the ovs suite to the features list so it will run as part of the CI

**Comment by oribon** [comment]
waiting for https://github.com/openshift/release/pull/18710, should allow all tests to pass.

**Pull request merged**